File: src/main/java/org/lvxnull/allowlist/AllowList.java
package org.lvxnull.allowlist;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.quiltmc.loader.api.ModContainer;
import org.quiltmc.loader.api.QuiltLoader;
import org.quiltmc.qsl.base.api.entrypoint.server.DedicatedServerModInitializer;
import org.quiltmc.qsl.command.api.CommandRegistrationCallback;
import org.quiltmc.qsl.lifecycle.api.event.ServerLifecycleEvents;

import java.io.IOException;
import java.io.UncheckedIOException;

public final class AllowList implements DedicatedServerModInitializer {
    public static final Logger LOGGER = LogManager.getLogger("Allowlist");
    public static final String MOD_ID = "allowlist";
    private static AllowListStorage storage;

    public static AllowListStorage getStorage() {
        return storage;
    }

    @Override
    public void onInitializeServer(ModContainer mod) {
        var meta = mod.metadata();
        var config = AllowListConfig.INSTANCE;

        try {
            storage = new AllowListStorage(QuiltLoader.getConfigDir().resolve(config.savePath()));
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }

        var commandProvider = new AllowListCommandProvider(meta, storage);

        ServerLifecycleEvents.STARTING.register(s -> {
            LOGGER.info("Starting allowlist {}", meta.version());
            try {
                storage.load();
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        });

        CommandRegistrationCallback.EVENT.register((dispatcher, ctx, env) -> commandProvider.register(dispatcher));

        ServerLifecycleEvents.STOPPING.register(s -> {
            LOGGER.info("Stopping allowlist {}", meta.version());
            try {
                storage.close();
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        });
    }
}

File: src/main/kotlin/org/lvxnull/allowlist/AllowList.kt
/*This file is part of allowlist.

  Allowlist is free software: you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software
  Foundation, either version 3 of the License, or (at your option) any later
  version.

  Allowlist is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
  more details.

  You should have received a copy of the GNU General Public License along with
  allowlist. If not, see <https://www.gnu.org/licenses/>. */

package org.lvxnull.allowlist

import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents
import net.fabricmc.loader.api.FabricLoader
import net.fabricmc.loader.api.metadata.ModMetadata
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import kotlin.reflect.jvm.isAccessible

@Suppress("UNUSED")
object AllowList: ModInitializer {
    val meta: ModMetadata = FabricLoader.getInstance().getModContainer("allowlist").get().metadata
    val logger: Logger = LogManager.getLogger("AllowList")

    val storage: AllowListStorage by lazy {
        AllowListStorage(FabricLoader.getInstance().configDir)
    }

    override fun onInitialize() {
        logger.info("Started allowlist ${meta.version}")

        ServerLifecycleEvents.SERVER_STOPPING.register {
            onClose()
        }
    }

    private fun onClose() {
        logger.info("Stopping allowlist ${meta.version}")
        if((::storage.apply { isAccessible = true }.getDelegate() as Lazy<*>).isInitialized()) {
            storage.close()
        }
    }
}
